src/Engine/Data/Texture.cpp

```diff
@@ -94,7 +98,92 @@ void Texture::bindImageTexture( int unit,
         uint( unit ), level, layered, layer, access, m_textureParameters.internalFormat );
 }
 
-void Texture::updateData( const void* data ) {
+void Texture::updateData( void* newData ) {
+    m_textureParameters.texels = newData;
+
+    // register gpu task to update opengl representation before next rendering
+    if ( m_updateDataTaskId.isInvalid() ) {
+        auto taskFunc = [this]() {
+            this->updateGL();
+            m_updateDataTaskId = Core::TaskQueue::TaskId::Invalid();
+        };
+        auto task          = std::make_unique<Core::FunctionTask>( taskFunc, getName() );
+        m_updateDataTaskId = RadiumEngine::getInstance()->addGpuTask( std::move( task ) );
+    }
```

Comment:
I still think this section should be a critical one. Two or more threads may ask for texture update simultaneously and the result should be unpredictable.
I propose (but have not tested nor verified) to have an instance member (a mutex) m_updateMutex and do the following to make any test_and_write on m_updateDataTaskId a critical section

```suggestion
    m_updateMutex.lock();
    if ( m_updateDataTaskId.isInvalid() ) {
        auto taskFunc = [this]() {
            this->updateGL();
            m_updateMutex.lock();
            m_updateDataTaskId = Core::TaskQueue::TaskId::Invalid();
            m_updateMutex.unlock();
        };
        auto task          = std::make_unique<Core::FunctionTask>( taskFunc, getName() );
        m_updateDataTaskId = RadiumEngine::getInstance()->addGpuTask( std::move( task ) );
    }
    m_updateMutex.unlock();
```

Reply:
Fine for me, I'll do it (with lock_guard). 
If I understand well, same shoud be done for setParameters, but I have a concern with non const ref acess with getParameters, since use code from different threads might gain write acess, so we may give access to the mutex (or to a lock) as well so that use code can lock the resource.